### PR TITLE
FTML optimizer implementation

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description ##
FTML optimizer implementation, requested in https://github.com/apache/incubator-mxnet/issues/9182

The default values of `beta1, beta2, epsilon` is the same with https://github.com/keras-team/keras-contrib/pull/110.

How to add test to verify the correctness of implementation? @sxjscience 

Here is the [test](https://github.com/keras-team/keras-contrib/blob/master/tests/keras_contrib/optimizers/ftml_test.py) for FTML in keras-contrib. Is that OK?

I have done only one experiment, FTML (val acc : 0.756210 at 10th epoch) can converge faster than momentum SGD (val acc : 0.684095 at 10th epoch) on cifar10, both using `lr = 0.001, wd = 0` and `resnet18_v1`. 

### Essentials ###
- [x] Passed code style checking (`make lint`)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: python/mxnet/optimizer.py [hunk 1]
```diff
@@ -529,6 +529,59 @@ def update_multi_precision(self, index, weight, grad, state):
...
+            grad = clip(grad, -self.clip_gradient, self.clip_gradient)
+        # get previous states
+        prev_d, prev_v, prev_z = state
+        # compute states
+        v_t = self.beta2 * prev_v + (1 - self.beta2) * square(grad)
+        d_t = (1 - pow(self.beta1, t)) / lr * (sqrt(v_t / (1 - pow(self.beta2, t))) + self.epsilon)
+        sigma_t = d_t - self.beta1 * prev_d
+        z_t = self.beta1 * prev_z + (1 - self.beta1) * grad - sigma_t * weight
+        # update weight
+        weight[:] = - z_t / d_t - lr * wd * weight
```
Comment (sxjscience) [inline python/mxnet/optimizer.py]: I think we should merge the wd term into the gradient. @szhengac could you help check this?

Comment (sxjscience) [reply python/mxnet/optimizer.py]: The rest formulas look good.

**Comment by sxjscience** [comment]
Testing the optimizer is a difficult problem and we haven't found a good solution. Currently I think this kind of test, which optimizes a simple problem and checks the error, should be enough, https://github.com/apache/incubator-mxnet/blob/master/tests/python/unittest/test_optimizer.py#L648-L672. Also, would you also add the C++ version? If C++ is added, we can test it against the python version.

**Comment by szhengac** [comment]
For weight decay, it may be correct, but for l2 regularizer, we can either incorporate the grad w.r.t. l2 regularizer into the complete grad or use following formula:
screen shot 2017-12-31 at 4 43 19 pm
where \lambda_2 is the regularization parameter. If elastic net is considered, the following one can be used:
screen shot 2017-12-31 at 9 46 48 am
where lambda_1 is the regularization parameter for $\ell_1$ part. 

Also, I think it is more efficient to update the powers of beta 1and beta 2 iteratively.

**Comment by ZiyueHuang** [comment]
Thanks for your comments.
- c++ version is added.
- I think weight decay is for l2 regularization in MXNet. l2 regularizer is now incorporated into the gradients.
- Right, it is more efficient to update the powers of beta1 and beta 2 iteratively. But `t = self._index_update_count[index]` which is not always strictly increasing from 0.

File: python/mxnet/optimizer.py [hunk 2]
```diff
@@ -529,6 +529,55 @@ def update_multi_precision(self, index, weight, grad, state):
         self._update_impl(index, weight, grad, state,
                           multi_precision=use_multi_precision)
 
+
+@register
+class FTML(Optimizer):
```
Comment (piiswrong) [inline python/mxnet/optimizer.py]: I thought we already have this

File: tests/python/unittest/test_optimizer.py [hunk 3]
```diff
@@ -333,6 +333,75 @@ def test_sparse_sgd():
...
+        assert(isinstance(grad, mx.nd.NDArray))
+        self._update_count(index)
+        lr = self._get_lr(index)
+        wd = self._get_wd(index)
+        t = self._index_update_count[index]
+
+        grad = grad * self.rescale_grad
+        if self.clip_gradient is not None:
+            grad = mx.nd.clip(grad, -self.clip_gradient, self.clip_gradient)
+        grad += wd * weight
```
Comment (sxjscience) [inline tests/python/unittest/test_optimizer.py]: We should clip after adding the gradient of L2. This is consistent with other optimizers.

File: src/operator/optimizer_op-inl.h [hunk 4]
```diff
@@ -485,6 +485,95 @@ inline void SGDMomUpdateEx(const nnvm::NodeAttrs& attrs,
...
+  template<typename DType>
+  MSHADOW_XINLINE static void Map(int i, DType* out, DType* weight, DType* grad,
+    DType* d, DType* v, DType* z, const DType lr, const DType beta1,
+    const DType beta2, const DType epsilon, const DType t,
+    const DType wd, const DType rescale_grad, const DType clip_grad,
+    const OpReqType req) {
+    using namespace mshadow_op;
+    const DType grad_i = clip_grad >= 0.0f
+        ? (clip::Map(rescale_grad * grad[i], clip_grad) + wd * weight[i])
+        : (rescale_grad * grad[i] + wd * weight[i]);
```
Comment (sxjscience) [inline src/operator/optimizer_op-inl.h]: We should clip after adding the gradient of L2. This is consistent with other optimizers.

File: python/mxnet/optimizer.py (hunk 2 above)
Comment (sxjscience) [reply python/mxnet/optimizer.py]: We have FTRL (Follow the regularized leader). This PR adds FTML (Follow the moving leader).

File: tests/python/unittest/test_optimizer.py (hunk 3 above)
Comment (ZiyueHuang) [reply tests/python/unittest/test_optimizer.py]: It seems that L2 term is out of clip in other optimizers, such as sgd in https://github.com/apache/incubator-mxnet/blob/master/src/operator/optimizer_op-inl.h#L76-L78.

Comment (sxjscience) [reply tests/python/unittest/test_optimizer.py]: In Adam, it’s clipped outside. So our current optimizes have such kinds of inconsist behavior. I think clip the gradient without we is wrong.

Comment (sxjscience) [reply tests/python/unittest/test_optimizer.py]: I mean without the WD part.

Comment (ZiyueHuang) [reply tests/python/unittest/test_optimizer.py]: Got it. Thanks. Now WD part is added into gradients.
